Warning for warning NU5128 when packastool with targetframework alias

---
**Issue moved from dotnet/sdk#13928**
- Please respond to @wli3.

---

_From @wli3 on Wednesday, September 30, 2020 9:03:56 PM_

Given
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>targetframeworkAlias</TargetFramework>
    <PackAsTool>true</PackAsTool>
    <OutputType>exe</OutputType>
  </PropertyGroup>

  <PropertyGroup Condition=" '$(TargetFramework)' == 'targetframeworkAlias' ">
    <TargetFrameworkIdentifier>.NETCoreApp</TargetFrameworkIdentifier>
    <TargetFrameworkVersion>v3.1</TargetFrameworkVersion>
    <TargetFrameworkMoniker>.NETCoreApp,Version=v3.1</TargetFrameworkMoniker>
  </PropertyGroup>

</Project>

```

run dotnet pack. There will be an extra necessary warning:
```
C:\work\sdk3\artifacts\bin\redist\Debug\dotnet\sdk\5.0.100-dev\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(207,5): warning NU5128: Some target frameworks declared in the dependencies group of the nuspec and the lib/ref folder do not have exact matches in the other location. Consult the list of actions below: [C:\work\temp\nugetalias\nugetalias.csproj]
C:\work\sdk3\artifacts\bin\redist\Debug\dotnet\sdk\5.0.100-dev\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(207,5): warning NU5128: - Add lib or ref assemblies for the netcoreapp3.1 target framework [C:\work\temp\nugetalias\nugetalias.csproj]
```

--- a/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs
+++ b/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs
@@ -234,9 +234,7 @@ namespace NuGet.Build.Tasks.Pack
                     new HashSet<NuGetFramework>(request.FrameworksWithSuppressedDependencies
                     .Select(t =>
                     {
-                        string translated = null;
-                        var succeeded = aliases.TryGetValue(t.Identity, out translated);
-                        if (succeeded)
+                        if (aliases.TryGetValue(t.Identity, out string translated))
                         {
                             return NuGetFramework.Parse(translated);
                         }

--- a/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs
+++ b/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs
@@ -219,13 +219,32 @@ namespace NuGet.Build.Tasks.Pack
                     .ToDictionary(msbuildItem => msbuildItem.Identity,
                     msbuildItem => msbuildItem.GetProperty("ProjectVersion"), PathUtility.GetStringComparerBasedOnOS());
             }
+
+            var aliases = new Dictionary<string, string>();
+            foreach (var tfm in assetsFile.PackageSpec.TargetFrameworks)
+            {
+                aliases[tfm.TargetAlias] = tfm.FrameworkName.GetShortFolderName();
+            }
+
             var nuGetFrameworkComparer = new NuGetFrameworkFullComparer();
             var frameworksWithSuppressedDependencies = new HashSet<NuGetFramework>(nuGetFrameworkComparer);
             if (request.FrameworksWithSuppressedDependencies != null && request.FrameworksWithSuppressedDependencies.Any())
             {
                 frameworksWithSuppressedDependencies =
                     new HashSet<NuGetFramework>(request.FrameworksWithSuppressedDependencies
-                    .Select(t => NuGetFramework.Parse(t.Identity)).ToList(), nuGetFrameworkComparer);
+                    .Select(t =>
+                    {
+                        string translated = null;
+                        var succeeded = aliases.TryGetValue(t.Identity, out translated);
+                        if (succeeded)
+                        {
+                            return NuGetFramework.Parse(translated);
+                        }
+                        else
+                        {
+                            return NuGetFramework.Parse(t.Identity);
+                        }
+                    }).ToList(), nuGetFrameworkComparer);
             }
 
             PopulateProjectAndPackageReferences(builder,

--- a/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs
+++ b/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs
@@ -74,8 +74,9 @@ namespace Dotnet.Integration.Test
         }
 
         [PlatformFact(Platform.Windows)]
-        public void PackCommand_PackToolUsingAlias_DoesNotWarn()
+        public void PackCommand_PackToolUsingAlias_DoesNotWarnAboutNoExactMatchInDependencyGroupAndLibRefDirectories()
         {
+            // Ref: https://github.com/NuGet/Home/issues/10097
             using (var testDirectory = msbuildFixture.CreateTestDirectory())
             {
                 // Arrange
@@ -103,6 +104,7 @@ namespace Dotnet.Integration.Test
                 msbuildFixture.RestoreProject(workingDirectory, projectName, string.Empty);
                 var result = msbuildFixture.PackProject(workingDirectory, projectName, $"-o {workingDirectory}");
 
+                // Assert
                 result.AllOutput.Should().NotContain("NU5128");
             }
         }

--- a/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs
+++ b/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs
@@ -74,6 +74,40 @@ namespace Dotnet.Integration.Test
         }
 
         [PlatformFact(Platform.Windows)]
+        public void PackCommand_PackToolUsingAlias_DoesNotWarn()
+        {
+            using (var testDirectory = msbuildFixture.CreateTestDirectory())
+            {
+                // Arrange
+                var projectName = "ConsoleApp1";
+                var workingDirectory = Path.Combine(testDirectory, projectName);
+                msbuildFixture.CreateDotnetNewProject(testDirectory.Path, projectName, " console");
+                var projectFile = Path.Combine(workingDirectory, $"{projectName}.csproj");
+                using (var stream = new FileStream(projectFile, FileMode.Open, FileAccess.ReadWrite))
+                {
+                    var xml = XDocument.Load(stream);
+
+                    ProjectFileUtils.AddProperty(xml, "PackAsTool", "true");
+                    ProjectFileUtils.ChangeProperty(xml, "TargetFramework", "myalias");
+
+                    var tfmProps = new Dictionary<string, string>();
+                    tfmProps["TargetFrameworkIdentifier"] = ".NETCoreApp";
+                    tfmProps["TargetFrameworkVersion"] = "v3.1";
+                    tfmProps["TargetFrameworkMoniker"] = ".NETCoreApp,Version=v3.1";
+                    ProjectFileUtils.AddProperties(xml, tfmProps, " '$(TargetFramework)' == 'myalias' ");
+
+                    ProjectFileUtils.WriteXmlToFile(xml, stream);
+                }
+
+                // Act
+                msbuildFixture.RestoreProject(workingDirectory, projectName, string.Empty);
+                var result = msbuildFixture.PackProject(workingDirectory, projectName, $"-o {workingDirectory}");
+
+                result.AllOutput.Should().NotContain("NU5128");
+            }
+        }
+
+        [PlatformFact(Platform.Windows)]
         public void PackCommand_PackNewDefaultProject_IncludeSymbolsWithSnupkg()
         {
             using (var testDirectory = msbuildFixture.CreateTestDirectory())

